search2/python/search/utils/beanstalkd.py

```diff
@@ -0,0 +1,199 @@
+def get_config_file_path(config_path_base_dir: str, assembly: str, suffix: str = ".y*ml"):
+    """Get config file path"""
+    paths = glob(path.join(config_path_base_dir, assembly + suffix))
+
+    if not paths:
+        raise ValueError(f"\n\nNo config path found for the assembly {assembly}. Exiting\n\n")
+
+    if len(paths) > 1:
+        print("\n\nMore than 1 config path found, choosing first")
+
+    return paths[0]
+
+def _specify_publisher(
+    submission_id, job_id, publisher_host, publisher_port, progress_event, event_queue
+):
+    return Publisher(
+            host = publisher_host,
+            port = publisher_port,
+            queue = event_queue,
+            message = Message(
+                event = progress_event,
+                queueID = job_id,
+                submissionID = submission_id,
+                data = None,
+            ))
+
+def listen(
+    handler_fn: Callable,
+    submit_msg_fn: Callable,
+    completed_msg_fn: Callable,
+    queue_conf: QueueConf,
+    tube: str,
+):
+    """Listen on a Beanstalkd channel, waiting for work.
+       When work is available call the work handler
+    """
+    hosts, ports = queue_conf.split_host_port()
+
+    for event in ("progress", "failed", "started", "completed"):
+        assert event in queue_conf.events
+
+    tube_conf = queue_conf.tubes[tube]
+    events_conf = queue_conf.events
+    clients = tuple(BeanstalkClient(host, port, socket_timeout=10) for (host, port) in zip(hosts, ports))
+
+    i = 0
+    while True:
+        i += 1
+        offset = i % len(hosts)
+        client  = clients[offset]
+
+        client.watch(tube_conf["submission"])
+
+        try:
+            job = client.reserve_job(5)
+        except BeanstalkError as err:
+            if err.message == BEANSTALK_ERR_TIMEOUT:
+                continue
+            raise err
+        try:
+            job_data = loads(job.job_data)
+
+            publisher: Publisher = _specify_publisher(
+                job_data['submissionID'],
+                job.job_id,
+                publisher_host=client.host,
+                publisher_port=client.port,
+                progress_event=events_conf["progress"],
+                event_queue=tube_conf["events"]
+            )
+
+            base_msg = {
+                "event": events_conf["started"],
+                "queueID": job.job_id,
+                "submissionID": job_data["submissionID"],
+            }
+
+            msg = submit_msg_fn(base_msg, job_data)
+
+            client.put_job_into(tube_conf["events"], dumps(msg))
+            res = asyncio.get_event_loop().run_until_complete(handler_fn(publisher, job_data))
+
+            base_msg["event"] = events_conf["completed"]
+            msg = completed_msg_fn(base_msg, job_data, res)
+
+            client.put_job_into(tube_conf["events"], dumps(msg))
+            client.delete_job(job.job_id)
+        except BeanstalkError as err:
+            print(f"Received error during execution: {err}")
```

Comment:
we should figure out a more comprehensive logging strategy at some point, because if you start the annotator with `pm2` afaict these print statemenets just fall into a void.